circle/src/domain.rs

```diff
@@ -0,0 +1,359 @@
+pub struct CircleDomain<F> {
+    // log_n corresponds to the size of the WHOLE domain
+    pub(crate) log_n: usize,
+    pub(crate) shift: Complex<F>,
+}
+
+impl<F: ComplexExtendable> CircleDomain<F> {
+    pub(crate) fn standard(log_n: usize) -> Self {
+        Self {
+            log_n,
+            shift: F::circle_two_adic_generator(log_n + 1),
+        }
+    }
+    fn is_standard(&self) -> bool {
+        self.shift == F::circle_two_adic_generator(self.log_n + 1)
+    }
+    pub(crate) fn points(&self) -> impl Iterator<Item = Complex<F>> {
+        let half_gen = F::circle_two_adic_generator(self.log_n - 1);
+        let coset0 = half_gen.shifted_powers(self.shift);
+        let coset1 = half_gen.shifted_powers(half_gen / self.shift);
+        coset0.interleave(coset1).take(1 << self.log_n)
+    }
+
+    #[instrument(skip_all, fields(log_n = %self.log_n))]
+    pub(crate) fn lagrange_basis<EF: ExtensionField<F>>(&self, point: Complex<EF>) -> Vec<EF> {
+        let domain = self.points().collect_vec();
+
+        // the denominator so that the lagrange basis is normalized to 1
+        // this depends only on domain, so should be precomputed
+        let lagrange_normalizer: Vec<F> = domain
+            .iter()
+            .map(|p| s_p_at_p(p.real(), p.imag(), self.log_n))
+            .collect();
+
+        let basis = domain
+            .into_iter()
+            .zip(&lagrange_normalizer)
+            .map(|(p, &ln)| {
+                // ext * base
+                v_0(p.conjugate().rotate(point)) * ln
+            })
+            .collect_vec();
+
+        batch_multiplicative_inverse(&basis)
+    }
+}
+
+impl<F: ComplexExtendable> PolynomialSpace for CircleDomain<F> {
+    type Val = F;
+
+    fn size(&self) -> usize {
+        1 << self.log_n
+    }
+
+    fn first_point(&self) -> Self::Val {
+        point_to_univariate(self.shift).unwrap()
+    }
+
+    fn next_point<Ext: ExtensionField<Self::Val>>(&self, x: Ext) -> Option<Ext> {
+        // Only in standard position do we have an algebraic expression to access the next point.
+        if self.is_standard() {
+            let gen = F::circle_two_adic_generator(self.log_n);
+            Some(point_to_univariate(gen.rotate(univariate_to_point(x))).unwrap())
+        } else {
+            None
+        }
+    }
+
+    fn create_disjoint_domain(&self, min_size: usize) -> Self {
+        let mut log_n = log2_ceil_usize(min_size);
+        // Right now we simply guarantee the domain is disjoint by returning a
+        // larger standard position coset, which is fine because we always ask for a larger
+        // domain. If we wanted good performance for a disjoint domain of the same size,
+        // we could change the shift. Also we could support nonstandard twin cosets.
+        assert!(
+            self.is_standard(),
+            "create_disjoint_domain not currently supported for nonstandard twin cosets"
+        );
+        while log_n == self.log_n {
```

Comment:
nit: it feels a bit weird to use a `while` loop to introduce a branching logic. Also, is it okay that `log_n < self.log_n` ?

Reply:
Yeah basically any standard domain where `log_n != self.log_n`, including smaller, is disjoint. I agree this is silly